File: questions/1653. Minimum Deletions to Make String Balanced/delete.cpp
class Solution {
public:
    int minimumDeletions(string s) {
        int n = s.size(), bcount = 0, res = INT_MAX;
        vector<int> A(n);
        A[n - 1] = (s[n - 1]=='a'?1:0);
        for(int i = n - 2; i >= 0; i--){
            A[i] = A[i + 1] + (s[i]=='a'?1:0);
        }
        for(int i = 0; i < n; i++){
            if(s[i] == 'b'){
                res = min(res, bcount + A[i]);
                bcount++;
            }
        }
        res = min(res, bcount);
        return res == INT_MAX?0:res;
    }
};

File: questions/1653. Minimum Deletions to Make String Balanced/delete.py
class Solution:
    def minimumDeletions(self, s: str) -> int:
        n = len(s)
        res = float('inf')
        bcount = 0
        A = [0]*n
        if s[n - 1] == 'a':
            A[n - 1] = 1
        for i in range(n - 2, -1, -1):
            A[i] = A[i + 1]
            if s[i] == 'a':
                A[i] += 1
        for i in range(0, n):
            if s[i] == 'b':
                res = min(res, bcount + A[i])
                bcount += 1
        res = min(bcount, res)
        if res == float('inf'):
            return 0
        return res
